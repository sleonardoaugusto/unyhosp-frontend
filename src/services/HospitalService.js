import apiClient from './config'

export default {
  get() {
    return apiClient.get('/hospitals/')
  },
  post(data) {
    return apiClient.post('/hospitals/', data)
  }
}