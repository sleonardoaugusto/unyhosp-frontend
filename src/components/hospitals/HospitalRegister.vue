<template>
  <div class="d-flex flex-wrap justify-content-center align-items-center flex-column">
    <div class="col-md-6 mb-3">
      <h2 class="type-h4 type-blue-smoke">Cadastrar Hospital</h2>
    </div>
    <form @submit.prevent="submit" class="col-md-6">
      <div class="form-group" :class="{ 'form-group--error': $v.hospital.name.$error }">
        <label class="form__label" for="HospitalName">Nome:</label>
        <input class="form__input" type="text" id="HospitalName" v-model.trim.lazy="$v.hospital.name.$model">
        <div v-if="!$v.hospital.name.required" class="error--message">Campo obrigat√≥rio</div>
      </div>
      <button class="button button-grey" type="submit">CADASTRAR</button>
    </form>
  </div>
</template>

<script>
import HospitalService from '@/services/HospitalService'
import { mapActions } from 'vuex'
import { required } from 'vuelidate/lib/validators'

export default {
  name: 'HospitalRegister',
  data: () => ({
    hospital: {
      name: ''
    }
  }),
  validations: {
    hospital: {
      name: { required }
    }
  },
  methods: {
    submit() {
      this.$v.$touch()
      if (!this.$v.$invalid) {
        HospitalService.post(this.hospital)
          .then(() => {
            const payload = { message: 'Cadastro realizado!', type: 'success' }
            this.add(payload)
          })
      }
    },
    ...mapActions({
      add: 'notification/add'
    })
  }
}
</script>

<style scoped>

</style>