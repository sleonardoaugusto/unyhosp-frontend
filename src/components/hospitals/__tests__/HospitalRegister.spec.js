import { shallowMount, createLocalVue } from '@vue/test-utils'
import HospitalRegister from '@/components/hospitals/HospitalRegister'
import Vuelidate from 'vuelidate'
import HospitalService from '@/services/HospitalService'

const localVue = createLocalVue()
localVue.use(Vuelidate)

describe('HospitalRegister.vue', () => {
  let wrapper

  beforeEach(() => {

    wrapper = shallowMount(HospitalRegister, {
      localVue
    })
  })

  it('component is rendering properly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('feedback form validation should appear', () => {
    const error = wrapper.find('.error--message')
    expect(error.text()).toBe('Campo obrigatório')
  })

  it('feedback form validation should NOT appear', () => {
    const data = { name: 'São Joaquim' }
    wrapper.setData({ hospital: data })
    wrapper.vm.$forceUpdate()
    expect(wrapper.findAll('.error--message').length).toBe(0)
  })

  it('click button should call submit method', () => {
    const submit = jest.fn()
    wrapper.setMethods({ submit })
    wrapper.find('form').trigger('submit')
    expect(submit).toHaveBeenCalled()
  })

  it('HospitalService post method should NOT be called on submit if data is INVALID', () => {
    const serviceSpy = jest.spyOn(HospitalService, 'post')
    wrapper.find('form').trigger('submit')
    expect(serviceSpy).not.toHaveBeenCalled()
  })

  it('HospitalService post method SHOULD be called on submit if data is VALID', () => {
    const data = { name: 'São Joaquim' }
    wrapper.setData({ hospital: data })
    const serviceSpy = jest.spyOn(HospitalService, 'post')
    wrapper.find('form').trigger('submit')
    expect(serviceSpy).toHaveBeenCalledWith(data)
  })
})