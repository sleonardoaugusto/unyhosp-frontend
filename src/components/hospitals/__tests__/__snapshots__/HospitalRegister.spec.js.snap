// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HospitalRegister.vue component is rendering properly 1`] = `
<div class="d-flex flex-wrap justify-content-center align-items-center flex-column">
  <div class="col-md-6 mb-3">
    <h2 class="type-h4 type-blue-smoke">Cadastrar Hospital</h2>
  </div>
  <form class="col-md-6">
    <div class="form-group"><label for="HospitalName" class="form__label">Nome:</label> <input type="text" id="HospitalName" class="form__input">
      <div class="error--message">Campo obrigat√≥rio</div>
    </div> <button type="submit" class="button button-grey">CADASTRAR</button>
  </form>
</div>
`;
